### PR TITLE
Roles and auth

### PR DESCRIPTION
**Pull request by JoshBroomberg**
This PR introduces roles and authorisation. 

Right now, these are only used in basic namespace level authorisation for the admin and application systems.

The basic design pattern:

**Controllers**
- Application controller includes pundit and handles all auth errors. It is the foundation layer of authorisation. 
- Each namespace has a top level controller (`apply_controller` and `admin_controller`). These controllers handle authentication and authorisation in before actions. Each one checks for the correct role on the correct user using pundit. 
- All sub controllers in the namespace inherit off the respective top level controller and thus authenticate and authorise before every action.
- the top level controller is responsible for routing within the namespace via its index action. This modularises routing and cleans up main index.

**Pundit**
- The pundit policy is called `NamespacePolicy` and defines an action for each relevant namespace that checks for the correct royce roles. 
- This design decision was simple. One policy per namespace was excessive because each namespace only has a single method. This would have resulted in `ApplyPolicy` with the action `can_apply?`. 
- Any more detailed authorisation can be done at a controller level using different policies later.

**Routing**
- All users land on /login and log in with facebook. Admins have their accounts tagged with the admin role once by someone with console access. We could include some kind of validation to ensure that the UID of anyone made admin matches a defined list to avoid human error.
- The main index now simply checks for the role of the correct user and directs them to the correct top level namespace controller. This top level controller handles further routing.
- All redirects from actions within a name sapce still go back to the main index. This is incase the user changes roles during an action and thus necessitate a change in namespace.

**Domains**
A discussion is required on whether we use subdomains. These domains complicate things somewhat because the route of the subdomain may not be the main root of the site. IE, you would want admin.ix.co.za to be a page, not a redundant routing action.

This means that routing between domains is annoying. I am in favour of hosting at journey.ixperience.co.za and then using /apply, /admin, /ascend, /explore, etc etc. I also think this is good UX.  Unless I am using an entirely different system I do not expect the url to change. Makes sense for google because each product is unique but not really for us.

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -36,6 +36,8 @@ gem "paperclip", "~> 4.3"
 gem "active_type", "~> 0.4"
 gem "time_splitter", "~> 1.2",
   git: "https://github.com/iExperience/time_splitter.git"
+gem "royce"
```
Comment (khanotations) [inline Gemfile]: Include versions

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -10,4 +12,9 @@ def model_error_messages(model)
   def render_json_error(msg, status = :unprocessable_entity)
     render json: msg, status: status
   end
+
+  def user_not_authorized
+    flash[:alert] = "You're in the wrong place, you've been taken back to safety."
```
Comment (khanotations) [inline app/controllers/application_controller.rb]: Nice message

File: app/policies/application_policy.rb [hunk 3]
```diff
@@ -0,0 +1,53 @@
+class ApplicationPolicy
```
Comment (khanotations) [inline app/policies/application_policy.rb]: Is this a template? sees excessive if not

File: db/migrate/20160809193619_create_royce.rb [hunk 4]
```diff
@@ -0,0 +1,22 @@
+class CreateRoyce < ActiveRecord::Migration
```
Comment (khanotations) [inline db/migrate/20160809193619_create_royce.rb]: I'm guessing this is also generated for you?

File: spec/policies/apply_policy_spec.rb [hunk 5]
```diff
@@ -0,0 +1,28 @@
+require 'rails_helper'
```
Comment (khanotations) [inline spec/policies/apply_policy_spec.rb]: Let's get rid of this

File: app/policies/application_policy.rb (hunk 3 above)
Comment (JoshBroomberg) [inline app/policies/application_policy.rb]: Yep, that's the default policy template that they suggest generating. You inherit off this for other policies and it makes set up much easier.

File: db/migrate/20160809193619_create_royce.rb (hunk 4 above)
Comment (JoshBroomberg) [inline db/migrate/20160809193619_create_royce.rb]: Indeed.

**Comment by JoshBroomberg** [comment]
Ok, fixes made.

**Comment by JoshBroomberg** [comment]
@khanotations Let me know what you think of the specs on this. Not sure the number of contexts is necessary.

File: Gemfile [hunk 6]
```diff
@@ -36,6 +36,8 @@ gem "paperclip", "~> 4.3"
 gem "active_type", "~> 0.4"
 gem "time_splitter", "~> 1.2",
   git: "https://github.com/iExperience/time_splitter.git"
+gem "royce", "~> 1.0.6"
+gem "pundit"
```
Comment (khanotations) [inline Gemfile]: no version here..?

**Comment by khanotations** [comment]
Looks good, but what are we going to do when a third role is added? Are we going to grow the contexts exponentially? How do we decide which combos we use?

**Comment by JoshBroomberg** [comment]
@khanotations fixes made. Let's discuss the role scenario when we get there. Answer is almost defintely no.

File: Gemfile [hunk 7]
```diff
@@ -9,7 +9,7 @@ gem "pg"
 gem "passenger" # Web server.
 gem "resque", "~> 1.22.0" # Background job processing.
 gem "flip" # Feature flags.
-gem "squeel"
+gem "squeel" # Advanced DSL for SQL generation.
```
Comment (khanotations) [inline Gemfile]: can we add version numbers to this, flip, passenger and all the gems. #mikeyang

**Comment by khanotations** [comment]
Please fix merge conflicts and add versions to gems. Then this looks good.

**Pull request merged**